internal/engine/target_queue.go

```diff
@@ -7,37 +7,113 @@ import (
+// Allows the caller to inject its own build strategy for dirty targets.
+type BuildHandler func(
+	target model.TargetSpec,
+	state store.BuildState,
+	depResults []store.BuildResult) (store.BuildResult, error)
+
+// A little data structure to help iterate through dirty targets in dependency order.
 type TargetQueue struct {
-	targets []model.TargetSpec
+	sortedTargets []model.TargetSpec
+
+	// The state from the previous build.
+	// Contains files-changed so we can do incremental builds.
+	state store.BuildStateSet
+
+	// The results of this build.
 	results store.BuildResultSet
+
+	// Whether the target itself needs a rebuilt, either because it has dirty files
+	// or has never been built before.
+	//
+	// A target with dirty files might be able to use the files changed
+	// since the previous result to build the next result.
+	needsOwnBuild map[model.TargetID]bool
+
+	// Whether the target depends transitively on something that needs rebuilding.
+	// A target that depends on a dirty target should never use its previous
+	// result to build the next result.
+	depsNeedBuild map[model.TargetID]bool
 }
 
-func NewImageTargetQueue(iTargets []model.ImageTarget) *TargetQueue {
+func NewImageTargetQueue(iTargets []model.ImageTarget, state store.BuildStateSet) (*TargetQueue, error) {
 	targets := make([]model.TargetSpec, 0, len(iTargets))
 	for _, iTarget := range iTargets {
 		targets = append(targets, iTarget)
 	}
 
-	return &TargetQueue{
-		targets: targets,
-		results: make(store.BuildResultSet, len(targets)),
+	sortedTargets, err := model.TopologicalSort(targets)
+	if err != nil {
+		return nil, err
+	}
+
+	needsOwnBuild := make(map[model.TargetID]bool)
+	for _, target := range sortedTargets {
+		id := target.ID()
+		if state[id].NeedsImageBuild() {
+			needsOwnBuild[id] = true
+		}
+	}
+
+	depsNeedBuild := make(map[model.TargetID]bool)
+	for _, target := range sortedTargets {
+		for _, depID := range target.DependencyIDs() {
+			if needsOwnBuild[depID] || depsNeedBuild[depID] {
+				depsNeedBuild[target.ID()] = true
+				break
+			}
+		}
 	}
+
+	results := make(store.BuildResultSet, len(targets))
+	return &TargetQueue{
+		sortedTargets: sortedTargets,
+		state:         state,
+		results:       results,
+		needsOwnBuild: needsOwnBuild,
+		depsNeedBuild: depsNeedBuild,
+	}, nil
 }
 
-// Checks if we have results for all the dependencies of this target.
-func (q *TargetQueue) isReady(target model.TargetSpec) bool {
-	for _, depID := range target.DependencyIDs() {
-		_, ok := q.results[depID]
-		if !ok {
-			return false
+func (q *TargetQueue) CountDirty() int {
+	result := 0
+	for _, target := range q.sortedTargets {
+		if q.needsOwnBuild[target.ID()] || q.depsNeedBuild[target.ID()] {
+			result++
 		}
 	}
-	return true
+	return result
 }
 
-func (q *TargetQueue) SetResult(id model.TargetID, result store.BuildResult) {
-	q.results[id] = result
+func (q *TargetQueue) BuildDirty(handler BuildHandler) error {
```

Comment:
This function name feels weird to me. It shouldn't only be used if the target has any dirty dependencies, it figured that out. Maybe:
`BuildWithState`?

I don't feel strongly though, it's a tough one to name.

Reply:
i changed it to RunBuilds

i'm still straightening out in my mind what this interface should look like. you can look at how this is currently handled in Bazel
https://github.com/bazelbuild/bazel/tree/1f684e1b87cd8881a0a4b33e86ba66743e32d674/src/main/java/com/google/devtools/build/lib/actions

where it has an ActionGraph class that returns "here's what you need to do to build this target" and an ActionGraphVisitor that does the visiting/ result accumulation

the way they do it is definitely way too many levels of abstraction for what we need, but it's a good place to look for inspiration